entgql/pagination.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2019-present Facebook
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Code generated by entc, DO NOT EDIT.
+
+package entgql
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	"golang.org/x/exp/slices"
+
+	"entgo.io/ent"
+	"entgo.io/ent/dialect/sql"
+	"github.com/vmihailenco/msgpack/v5"
+)
+
+// OrderDirection defines the directions in which to order a list of items.
+type OrderDirection string
+
+const (
+	// OrderDirectionAsc specifies an ascending order.
+	OrderDirectionAsc OrderDirection = "ASC"
+	// OrderDirectionDesc specifies a descending order.
+	OrderDirectionDesc OrderDirection = "DESC"
+)
+
+// Validate the order direction value.
+func (o OrderDirection) Validate() error {
+	if o != OrderDirectionAsc && o != OrderDirectionDesc {
+		return fmt.Errorf("%s is not a valid OrderDirection", o)
+	}
+	return nil
+}
+
+// String implements fmt.Stringer interface.
+func (o OrderDirection) String() string {
+	return string(o)
+}
+
+// MarshalGQL implements graphql.Marshaler interface.
+func (o OrderDirection) MarshalGQL(w io.Writer) {
+	io.WriteString(w, strconv.Quote(o.String()))
+}
+
+// UnmarshalGQL implements graphql.Unmarshaler interface.
+func (o *OrderDirection) UnmarshalGQL(val interface{}) error {
+	str, ok := val.(string)
+	if !ok {
+		return fmt.Errorf("order direction %T must be a string", val)
+	}
+	*o = OrderDirection(str)
+	return o.Validate()
+}
+
+// Reverse the direction.
+func (o OrderDirection) Reverse() OrderDirection {
+	if o == OrderDirectionDesc {
+		return OrderDirectionAsc
+	}
+	return OrderDirectionDesc
+}
+
+// PageInfo of a connection type.
+type PageInfo[T any] struct {
+	HasNextPage     bool       `json:"hasNextPage"`
+	HasPreviousPage bool       `json:"hasPreviousPage"`
+	StartCursor     *Cursor[T] `json:"startCursor"`
+	EndCursor       *Cursor[T] `json:"endCursor"`
```

Comment:
```suggestion
	StartCursor     T `json:"startCursor"`
	EndCursor       T `json:"endCursor"`
```

Reply:
I wanted to emphasis that this is limited to `Cursor`s. Don't you think so?